third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -172,32 +340,196 @@ struct LowerInstructionSchedHints
   }
 };
 
-struct InsertInstructionSchedHints
-    : public triton::impl::InsertInstructionSchedHintsBase<
-          InsertInstructionSchedHints> {
+struct TritonAMDGPUInsertInstructionSchedHints
+    : public triton::impl::TritonAMDGPUInsertInstructionSchedHintsBase<
+          TritonAMDGPUInsertInstructionSchedHints> {
   void runOnOperation() override {
     MLIRContext *ctx = &getContext();
     ModuleOp mod = getOperation();
 
-    mod->walk([ctx](triton::DotOp dot) {
-      if (dyn_cast<mlir::scf::ForOp>(dot->getParentOp())) {
+    mod.walk([this, ctx](scf::ForOp forOp) {
+      triton::DotOp dot = nullptr;
+      size_t dotCounter = 0;
+      forOp->walk([&dot, &dotCounter](triton::DotOp op) {
+        dot = op;
+        ++dotCounter;
+      });
+      // Note, instruction schedule barriers are inserted only in the case of
+      // a single `tt.dot` op in a `scf::ForOp` scope in the current
+      // implementation.
+      if (dotCounter == 1) {
         mlir::OpBuilder rewriter(ctx);
         rewriter.setInsertionPointAfter(dot);
         rewriter.create<triton::amdgpu::InstructionSchedHint>(dot->getLoc());
+        annotateDotUsageOnLoadStore(forOp);
       }
     });
   }
+
+  template <typename Type>
+  llvm::SmallVector<Type> getUsersOfType(Value value) const {
+    llvm::SmallVector<Type> concreteUsers;
+    for (auto user : value.getUsers()) {
+      if (auto concreteUser = llvm::dyn_cast<Type>(user))
+        concreteUsers.push_back(concreteUser);
+    }
+    return concreteUsers;
+  }
+
+  // Go through a single use chain to get the final value after all unary
+  // ops - e.g., `convert_layout`, `fp_to_fp`, etc.
+  FailureOr<Value> rewindUnaryOps(Value value) const {
+    auto findUnaryOps = [](Value value) {
+      llvm::SmallVector<Operation *> unaryOps;
+      for (Operation *op : value.getUsers()) {
+        if (op->getNumOperands() == 1)
+          unaryOps.push_back(op);
+      }
+      return unaryOps;
+    };
+    auto unaryOps = findUnaryOps(value);
+    if (!unaryOps.empty()) {
+      if (unaryOps.size() != 1)
+        return failure();
+      value = unaryOps[0]->getResult(0);
+      unaryOps = findUnaryOps(value);
```

Comment:
Sorry if this is a dumb comment, but should  you not call `rewindUnaryOps` here? I mean what happens if you have `finalValue = UnaryOp(unaryOp(unaryOp(unaryOp(value))))`?